rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,767 @@
+  Proxy-->>Client: SSH session established
+```
+
+#### Leaf Cluster Flow
+
+```mermaid
+---
+title: SSH MFA Connection Flow (Leaf Cluster)
+---
+sequenceDiagram
+  autoNumber
+
+  participant Client
+  participant rMFA as Local MFA Service
+  participant Proxy as Leaf Proxy Service
+  participant Dec as Leaf Decision Service
+  participant lMFA as Leaf MFA Service
+  participant SSH as Leaf SSH Service
+
+  participant Host as Target SSH Host
+
+  Client->>Proxy: Dial SSH
+  Proxy->>Dec: EvaluateSSHAccess
+  Dec-->>Proxy: Permit
+  Proxy->>SSH: Proxy SSH connection (stapled permit)
+  SSH->>SSH: Is MFA required?
+
+  alt MFA required
+    SSH->>Client: Keyboard-interactive (MFA required)
+    Client->>rMFA: CreateChallenge (SIP)
+    rMFA-->>rMFA: Generate and store MFA challenge with SIP
+    rMFA-->>Client: MFA challenge
+
+    Client->>Client: Solve MFA challenge
+    Client->>rMFA: ValidateChallenge (MFA response)
+    rMFA-->>rMFA: Validate MFA response
+    rMFA-->>lMFA: CreateValidatedChallenge (SIP)
+    rMFA-->>Client: Challenge validated
+
+    Client->>SSH: Keyboard-interactive (challenge name)
+    SSH->>lMFA: GetValidatedChallenge (challenge name)
+    lMFA-->>SSH: ValidatedChallengeResponse and SIP
+    SSH-->>SSH: Compute SIP
+    SSH-->>SSH: Verify SIP matches
+  end
+
+  SSH->>Host: Establish SSH connection
+  Host-->>SSH: SSH connection established
+  SSH-->>Proxy: SSH connection established
+  Proxy-->>Client: SSH session established
+```
+
+### Security
+
+In addition to the risks raised in [Access Control Decision API (RFD
+0024e)](https://github.com/gravitational/Teleport.e/blob/master/rfd/0024e-access-control-decision-api.md), there are a
+few risks specific to this RFD.
+
+#### Replay Attacks Risk
+
+MFA challenge responses could potentially be captured and replayed by an attacker to gain unauthorized access. Although
+this risk has already existed using per-session MFA SSH certificates, it is important to address it in the new design as
+well.
+
+Mitigations:
+
+1. MFA challenges and their responses will be single-use only. Once a challenge has been successfully validated, it
+   cannot be reused.
+1. The MFA challenge is time-bound and will expire after a duration (e.g., 5 minutes).
+1. Each MFA challenge will include a SIP that binds the challenge to a specific user session.
+
+#### New RPCs Attack Surface Risk
+
+This RFD introduces an MFA service which exposes four RPCs: `CreateChallenge`, `ValidateChallenge`,
+`CreateValidatedChallenge` and `GetValidatedChallenge`. These could potentially be exploited by an attacker to DoS the
+service by flooding it with requests.
+
+Mitigations:
+
+1. Only authenticated end user clients are authorized to call the `CreateChallenge` and `ValidateChallenge` RPCs,
+   requests from other sources will be rejected.
+1. Only the Teleport Proxy is authorized to call the `CreateValidatedChallenge` RPC, requests from other sources will be
+   rejected.
```

Comment:
`CreateValidatedChallenge` should check the name of the cluster that the request is being created for, and the challenges should probably be stored separately for each cluster.